### ISSUE TITLE
Use fscache on Windows

### ISSUE THREAD

**Issue by mehrdadn** [libgit2/libgit2#5115]
Hi!

Currently libgit2 (which affects users of e.g. TortoiseGit) is extremely slow on Windows due to the differing performance characteristics of Linux and Windows file systems.

Git-for-Windows has [solved this problem by replacing functions such as `lstat`](https://github.com/git-for-windows/git/blob/6dad4f177b65ab33185f8e018c275e5c3fadb597/compat/win32/fscache.c#L385) with versions that return cached results from directory enumerations, and by enabling and disabling the cache at appropriate times during which the file system is not modified.

I expected libgit2 to be using this code, but it appears that it is not. I think it would be great if it could take a similar approach (perhaps by using the code there) to improve performance on Windows.

**Comment by ethomson** [comment]
Yep, so do we.

If your issue is that tortoisegit is slow, please either raise issues with them so that they can investigate and raise actionable, reproducible issues with this project that could be investigated and fixed.  Or identify these hotspots yourself.

Because tortoisegit being slow could be tortoisegit’s code, git’s code, our code or some interaction of the 3.

**Comment by mehrdadn** [comment]
Oh, I've already identified the problem is in this project. Here's one example hotspot:

	KERNELBASE.dll	GetFileAttributesExW
	libgit2_tgit.dll	lstat_w + 0x35	 libgit2\src\win32\posix_w32.c(305)
	libgit2_tgit.dll	p_stat + 0x3f	 libgit2\src\win32\posix_w32.c(724)
	libgit2_tgit.dll	git_futils_readbuffer_updated + 0x5a	 libgit2\src\fileops.c(184)
	libgit2_tgit.dll	config_is_modified + 0x49	 libgit2\src\config_file.c(143)
	libgit2_tgit.dll	config_refresh + 0x3d	 libgit2\src\config_file.c(176)
	libgit2_tgit.dll	config_get + 0x23	 libgit2\src\config_file.c(302)
	libgit2_tgit.dll	get_entry + 0x92	 libgit2\src\config.c(739)
	libgit2_tgit.dll	git_config__lookup_entry + 0x1a	 libgit2\src\config.c(772)
	libgit2_tgit.dll	git_config__cvar + 0x33	 libgit2\src\config_cache.c(94)
	libgit2_tgit.dll	git_repository__cvar + 0x44	 libgit2\src\config_cache.c(116)
	libgit2_tgit.dll	git_index_set_caps + 0x94	 libgit2\src\index.c(574)
	libgit2_tgit.dll	git_repository_index__weakptr + 0xa9	 libgit2\src\repository.c(1186)
	libgit2_tgit.dll	attr_setup + 0x148	 libgit2\src\attr.c(347)
	libgit2_tgit.dll	collect_attr_files + 0x7a	 libgit2\src\attr.c(498)
	libgit2_tgit.dll	git_attr_get_many_with_session + 0xba	 libgit2\src\attr.c(131)
	libgit2_tgit.dll	filter_list_check_attributes + 0x82	 libgit2\src\filter.c(434)
	libgit2_tgit.dll	git_filter_list__load_ext + 0x11b	 libgit2\src\filter.c(521)
	libgit2_tgit.dll	git_filter_list_load + 0x3a	 libgit2\src\filter.c(586)
	libgit2_tgit.dll	git_repository_hashfile + 0xae	 libgit2\src\repository.c(2563)
	TortoiseGit.dll	CGitIndexList::GetFileStatus + 0x2d3	 tortoisegit\src\git\gitindex.cpp

This shouldn't be surprising or unexpected since as far as I can tell this library makes no attempt to use something like an fscache at all whatsoever.

**Comment by ethomson** [comment]
> This shouldn't be surprising or unexpected since as far as I can tell this library makes no attempt to use something like an fscache at all whatsoever.

Again, yes, it does.  In fact there are two different caches that are in your stack trace.

Are we perfect?  No.  Is their room for improvement?  Always.  But again, we need _actionable_, _reproducible_ steps to identify what you're seeing.  "It's slow" is neither of those things, and a stack trace isn't much more helpful.

If TortoiseGit is calling `GetFileStatus` over and over again, then yes, that would be slow.  We're not caching anything beyond invocations to a command, **exactly like Git isn't caching anything beyond invocations to `git status`**.

If you want to help make this particular call fast, then we can do that.  It's going to involve **working with TortoiseGit** to design an API that is re-entrant and keeps the caches alive through multiple invocations.

An issue that says "libgit2 is slow" isn't the way to get there.

**Comment by mehrdadn** [comment]
I can't really give reproducible steps because I'm not a direct user of your library.

However, what I don't understand is, aren't the stack traces above enough to show that file system caching is not occurring even for a single call? I've shown 2 different invocations of `GetFileAttributesExW` for the exact same file, both as a result of `GetFileStatus`... and the file system isn't being modified in the process, so clearly if it were caching it would be reducing those down to 1. It seems pretty actionable and rather clearly not the fault of TortoiseGit? I'm confused how you say there's caching here when it's hitting the file system despite those.

**Comment by ethomson** [comment]
> However, what I don't understand is, aren't the stack traces above enough to show that file system caching is not occurring even for a single call?

It's not a single call.  It's _two_ calls.  We don't cache between two different calls, because there's an arbitrary amount of time between them.  We need to _inspect what changed_ in that time.

>  and the file system isn't being modified in the process, so clearly if it were caching it would be reducing those down to 1.

No, there's no caching that exists _between multiple calls_, nor should there be.  When would we invalidate this cache and hit the filesystem again?  How would we determine when this cache should be dirtied?

I think that you may have an overly optimistic idea of what `core.fscache` does.  It is _not_ some thing that keeps status cached through multiple invocations to git.  It creates a cache for _a single invocation of a git executable_.  If you call `git status`, it will read the filesystem information, cache that, and then use that cache _throughout the execution of the `git status` call.  This is because calling `lstat` repeatedly is slow on Windows.  However, this cache is - of course - destroyed as soon as the process exits.  So if you call `git status` a second time, there's no cache to use.  So it hits the filesystem again.

_Exactly as we do between calls to `git status`_.

If TortoiseGit is calling status repeatedly, we're going to hit the filesystem.  That's a truism.  Our goal is to reduce the amount of filesystem calls and work that happens _in that function_.  But, yes, any subsequent calls to `git_status` **are going to hit the filesystem** by definition.  Exactly like git does.

Either TortoiseGit should cache the results of `git_status`, find another way to reduce the number of calls to the function, or work with us to create a version of the function that is re-entrant and holds on to caches.

**Comment by mehrdadn** [comment]
> It's not a single call. It's *two* calls.

**Comment by mehrdadn** [comment]
> It's not a single call. It's *two* calls.

No, it's **not** two calls. It's **ONE** call. Have you tried it at all? Here:

	// Usage:  echo foo > a && git add a && git commit -m "Initial commit" && test.exe a
	#include <git2.h>
	#include <stdio.h>

	int main(int argc, char *argv[]) {
		git_libgit2_init();
		git_repository *repo = NULL;
		int error = 0;
		char const *repo_path = ".git";
		if ((error = git_repository_open(&repo, repo_path)) == 0) {
			fprintf(stderr, "git_repository_open(\"%s\") succeeded\n", repo_path);
			fprintf(stderr, "Press ENTER to continue..."); getchar();
			for (int i = 1; i < argc; ++i) {
				git_oid oid;
				if (git_repository_hashfile(&oid, repo, argv[i], GIT_OBJECT_BLOB, NULL) == 0) {
					fprintf(stderr, "git_repository_hashfile(%s) succeeded\n", argv[i]);
				} else {
					fprintf(stderr, "git_repository_hashfile(%s) failed\n", argv[i]);
				}
			}
			fprintf(stderr, "Finished; press ENTER to quit..."); getchar();
			git_repository_free(repo);
		} else {
			fprintf(stderr, "git_repository_open(\"%s\") returned: %d\n", repo_path, error);
		}
		return error;
	}

Now run ProcMon between the two ENTER key presses and watch the screen get **FLOODED** with multiple FS hits for several files; here's just a snippet:

![](https://i.imgur.com/LY2sJhP.png)

**Comment by mehrdadn** [comment]
@ethomson: Bump?

**Comment by ethomson** [comment]
> No, it's not two calls. It's ONE call. Have you tried it at all?

No; you didn't give me anything to try until the `git_repository_hashfile` example, which is indeed helpful.

If `git_repository_hashfile` is a pain point for your application then a new issue with the repro steps that you provided would be helpful.  This isn't something that I can tackle in the near future but perhaps somebody else can investigate.

If you'd like to work on this yourself, we can iterate on a PR or discuss on slack or irc.

Thanks!

**Comment by mehrdadn** [comment]
"My application"? Like I said in the beginning, I don't *have* an application. It's a pain point for **TortoiseGit**. Which you kept blaming.

I had already spent **hours** of my time narrowing down the problem to libgit2. I reported the issue and explained it clearly. You immediately denied it without any investigation, repeatedly told me I'm wrong and persisted in making the bogus claim that these are multiple calls by TortoiseGit, and closed the PR in my face. I spent 2 **more** hours learning how to use the library just to prove to you that contrary to your insistence the problem has absolutely *nothing* to do with TortoiseGit.

It would've taken you (or even me, now that I *do* know the API) just 5-10 minutes to verify that I'm correct. Yet you didn't put in any effort and just fabricated things to tell me I'm wrong. And now that you realize, now you'd like me to help fix it too? Not even a simple apology or even reopening the issue now that you realize it was valid and you closed it incorrectly? Sorry, I prefer to cooperate, not fight. I was only trying to help this project and its users, but I don't particularly enjoy getting gaslighted and spending 20x more of my time because the person I'm trying to help is uncooperative. Makes me regret spending so much time on this.

**Comment by ethomson** [comment]
Indeed, I was not blaming TortoiseGit.  I suggested that they were the best people to help analyze the problems that they're seeing with their application, since they actually know the APIs that they're calling and how.

Your example application, above, appears to read the `index` and `config` several times.  That seems both unnecessary and solvable!  So let's assume that we optimize for the case where somebody opens a repository, calls `git_repository_hashfile` on a file and then disposes the repository.  But will that actually benefit **TortoiseGit**?  Only if they actually perform this open / hashfile / close pattern, and I suspect that they are not.

I'm not trying to be uncooperative, I'm trying to get more information and have our bug list contain bugs that have reproducible steps.  And this *particular* issue proposes a solution; it doesn't document a problem.

We're happy to chase perf issues, but we need to know what they are.

**Comment by mehrdadn** [comment]
You haven't even attempted to understand the example I gave you.

> open / hashfile / close pattern

Again, no, this is **NOT** what is happening. You *still* fabricate things out of thin air.

Read my example again:

>> Now run ProcMon **between** the two **ENTER key** presses 

What happens **between** the key presses is **only** a `git_repository_hashfile` call.

I'm tired of this.

**Comment by ethomson** [comment]
> You haven't even attempted to understand the example I gave you.

> The flooding that happens between the key presses is only for a single git_repository_hashfile call.

Yes, I understand that.  My point is: is this something that TortoiseGit actually _does_?  If not, then why should we go fix it?  You will get literally _no_ performance gain?

**Comment by mehrdadn** [comment]
What?? TortoiseGit is calling `git_repository_hashfile`, and **every call** to it is hitting the file system like a dozen times. You can easily reduce that by a factor of 2x or 3x. How can reducing FS hits by  **not** result in a performance gain??  Opening/closing just increases the delay, not decreases it??

**Comment by ethomson** [comment]
> I'm tired of this.

I am, too.  So let's stop for a minute, okay?

I think that you have - absolutely - the best intentions in opening this issue and that you want TortoiseGit to improve (and libgit2 to improve in order to enable that).

**I want that, too**.

You think that there are performance issues in libgit2.

**I think that, too**.

We want the same things here, but we seem to have gotten off on the wrong foot.  And that's regrettable.  I apologize if I came off as uncooperative.  I propose that we hit reset here.

**Comment by mehrdadn** [comment]
Okay, I'm happy to reset, but only after you put some substantial time into this too. At the moment I've spend several hours of my time and I don't think it'd be fair or a great use of my time to keep doing that (even after a reset), for several reasons: (1) this is neither my project nor am I desperate for a solution, (2) it's your project and I'm just trying to help other people, and (3) you're much more familiar with your library and git internal than I am, and for that (and more) reasons it would continue to take me far too disproportionately long to do so much of these than it would take you.

**Comment by mehrdadn** [comment]
For what it's worth, after all my investigations (and my general feel/experience with Windows FS performance characteristics, which I can't really put into words here, as well as tons of ProcMon traces from TortoiseGit, etc.) I have every reason in the world to believe libgit2 is bottlenecking TortoiseGit. If you somehow find actual evidence that this is not the case, I'll be happy to help investigate things on that side, which is I think where I could be reasonably effective. Otherwise, I don't see TortoiseGit as having any relevance anymore, so we should focus on libgit2. And for what it's worth—if this was my project, I would be seeking to reduce those calls _regardless_ of whether it was directly impacting any end user at all, because they're just *so* bad on their own. Both in principle (race conditions) and in practice—Windows FS calls are so costly that those *will* be problems for the next guy and that could well prevent people from even _trying_ to build faster tools with this.

So once we move on to libgit2, I might be able to help if you'd like me to benchmark/test things out or figure out how to do something. Or even other stuff... like if you're not sure where an issue is, I'm also happy to try to help. But only *after* you let me know what you already tried and where you failed (or where you genuinely think I might be more efficient), because I honestly can't bear the primary responsibility of those and simply can't invest anything remotely close to the majority of the effort of this on my own.

**Comment by carlosmn** [comment]
If you are not a direct user of the library, the best direction really is to go through the application where you are feeling the pain. We don't know which version they're using or whether they have made changes to the version they are using.

We also do not know how they use the library or whether a single function call really is at the bottom of the issues you are seeing in the application or whether it's the way they are using it or whether it's because they are calling it too often and some cache on their side would be a better option for them. Have you talked to the TortoiseGit developers about the issues you are seeing? Without their input most of the time it's going to be rather hard to actually pinpoint what the issue is.

You have pointed out an issue with `git_repository_hashfile` and it does look like it keeps trying to open the index file. Having a small piece of code that shows an issue is very helpful thank you. This looks like it might be about some concurrency/safety changes where this function keeps using the index and configuration accessors in an old way assuming it's cached whereas we're returning a brand new copy of the index. If that is indeed the case, this might be easy to fix, but it also will have nothing to do with `fscache` or cached results in `lstat` calls.

If the issue is multiple configuration and index reads due to calling the method multiple times, then it's not about filesystem metadata caching, but the fact that we cannot know a priori whether the index file or the configuration have changed without looking at them. In this case git has an easier time of things, since it knows it runs for a very limited amount of time and it can cache things in globals. This is not something that is possible for libgit2. It might be useful for a caller to specify when they're starting a complex short-lived operation that allows libgit2 to ignore any configuration or index changes that might happen. But that's going to require figuring out what callers (i.e. an application like TortoiseGit) are doing and what kind of API would work for them.

As Ed pointed out, this isn't necessarily something we can solve on our own, even if the one function could be improved, libgit2 does not know what you are doing at a higher level, so it cannot know when it can avoid loading changing data. This is the same for reloading the index and cofniguration as for reloading filesystem caches (if that is even involved in whatever issues you are seeing). This project is a library and cannot know what the larger objective is. This is why it's been mentioned several times that you should start talking to the application you're having issues with. They're the ones who would know how that would work and what kind of API they might need.

And if I might point something for future interactions, not just for this project but in general, the issue as reported started off assuming that we do not perform filesystem metadata caching and proposing we copy what Git for Windows does. This arrival with a solution to a nebulous problem is not something that  project contributors can really do something with, particularly since, as it's been pointed at, there is already some caching going on at the layer you pinpointed. It was only later that you provided something that is something that definitely runs within our project and as I have tried to explain, even we might not be able to fix an issue that happens at a higher level. The project contributors which you are asking to fix an issue you are experiencing are often overworked and it really helps for the asker to lay down what they are seeing and the results of their experiments from the start, instead of telling them what the solution should be.

**Comment by mehrdadn** [comment]
> If you are not a direct user of the library, the best direction really is to go through the application where you are feeling the pain.

Don't worry about my pain. My pains are irrelevant to your project. Well, except for the fact that the only pain I'm feeling is on this thread. I've only been trying to help you and other people. That's all. My own pains are quite secondary. I would care about this just as much as I do now if you gave me an infinitely fast machine.

> We also do not know how they use the library or whether a single function call really is at the bottom of the issues you are seeing in the application or whether it's the way they are using it or whether it's because they are calling it too often and some cache on their side would be a better option for them.

First, you do **not** need to know this at all. I was not coming here seeking a diagnosis or advice for what to do on my end, nor was I asking you to make TortoiseGit faster or to solve my pains or TortoiseGit's. I was just trying to help you see a serious performance problem that stands independently of any of those and hoping its significance would be glaringly obvious regardless of whose software I personally may or may not be using. If it isn't, that's the end of the story for me.

Second, even if TortoiseGit were to hypothetically call `git_repository_hashfile` in *the most abysmally inefficient way possible*, it is still a clear fact that every call is doing something like 3x as many FS hits as it needs to. That means you're still going to speed it up considerably by fixing this. And not just for TortoiseGit, but for every other user of your library who would call things optimally, now or in the future.

> This project is a library and cannot know what the larger objective is.

I assume this library was not an attempt at abstract math, but with some general types of use cases in mind. To the extent that you don't know the exact specifics of a given use case, it should be *glaringly obvious* at this point that your library would cause significant slowdowns for clients *on top* of any slowness they cause. So if you are convinced the C code is proof that there is a significant problem worth fixing in this library independently of any downstream users, then we can work together to resolve it. If not, then my cause is already lost, and any personal pain points I might have with TortoiseGit aren't relevant to this project.

> The project contributors which you are asking to fix an issue you are experiencing are often overworked

As if I'm underworked or somehow have more responsibility for this? Again: I did not post here to solve my personal problem. So anything more you ask me to do can't be based on the premise of *"you're asking us for something and we're overworked, so sorry, you'll need to put in most of the effort yourself"*, but rather *"we both want to help other people and are both sacrificing our time for them, so how can we do that more efficiently?"* Because, you have to realize: it is obviously true that I would benefit if this were fixed, but *if I thought this issue only affected me, I would not have posted at all in the first place*.

> And if I might point something for future interactions, not just for this project but in general, the issue as reported started off assuming that we do not perform filesystem metadata caching

Which was well-supported by the evidence I've shown, and the opposite was disputed by the same evidence.

> and proposing we copy what Git for Windows does. [...]

Yes, because generally I try to propose solutions if I have them, not just go to people with problems. But you could've ignored the solutions if you'd have liked. Where I went wrong was wrongly assuming that when you saw my stack trace, you'd grab a couple traces yourself, immediately verify the multiple calls (it was literally impossible to miss), sand then wonder how to solve the problem.

> It was only later that you provided something that is something that definitely runs within our project

No, it was earlier than that, but after the initial post on both sides. My *initial* post didn't have any evidence, and consequently it was fine that you guys were skeptical. But that's where the legitimate ambiguity ended. After that, I made it quite clear the caching is almost certainly not happening *before* I ever provided any C. Specifically, in my *very first* reply, I already made it very clear with actual concrete stack traces that TortoiseGit is calling the *same* API but that that *one* function is hitting the FS multiple times. That's admittedly not 100% mathematically airtight proof to execute someone based on, but it's *pretty darn good evidence* that maybe the caching you claim isn't happening, and that you should investigate on your end to see if the claim you're making is actually true. To simply reject it *again* without testing anything at all or providing *any* evidence whatsoever to the contrary, and to simply assert outright that this is a result of *multiple (!) calls* when *all* the evidence points to it being the exact same call, was quite ridiculous.

**Comment by tiennou** [comment]
The only caller of `git_repository_hashfile` in TortoiseGit is in [GetFileStatus](https://github.com/TortoiseGit/TortoiseGit/blob/d7bf7ad68a5c9acf6732063f2f7558f60452dfbc/src/Git/GitIndex.cpp#L225). The only "outside" caller of that (after some method overriding), is [CCachedDirectory](https://github.com/TortoiseGit/TortoiseGit/blob/0ed9575a5652a9b0a4ed0d90b79469779d36a518/src/TGitCache/CachedDirectory.cpp#L380). There's also this [comment](https://github.com/TortoiseGit/TortoiseGit/blob/febae42331d9f071bc5324cf55e2e00df2de843b/src/Git/GitFolderStatus.h#L38), which seems to document ho that caching.

We're thus looking at another cache, one layer above us, and GetFileStatus seems to be part of its status-caching-update mechanism. As we pointed out, this code makes no obvious sense to us, and it must be explained why it exists, and what we could provide in the library to make it perform better. The only thing I can tell is that TortoiseGit is *explicitly* asking us to read/filter the file, updating the odb/index to match what's on disk, because that's what `git_repository_hashfile` is meant to [do](https://libgit2.org/libgit2/#HEAD/group/repository/git_repository_hashfile). So the first (simple) thing I'd consider for fixing performance is to assess whether applying filters makes any sense for the use case TortoiseGit has — as both your FS traces are about loading the filters from config, then loading the attributes for the filter rules — but that must be checked against *what TortoiseGit uses that function for*, which we're clearly not in a position to do (or even remotely interested, I'm the macOS developer). A more complete solution would entail determining *why TortoiseGit isn't using the [git_status_list](https://libgit2.org/libgit2/#HEAD/group/status/git_status_list_new) API*, which is designed for that kind of stuff.

I'm not trying to deflect blame for the performance characteristics of `git_repository_hashfile`. I hope that this (kind of|hopefully) technical assessment will be taken as a positive step toward fixing a performance issue that sure looks painful, and to reiterate that we're interested in a PR for those kind of fixes (but turning our "quick repository hashing function" into "a better form of git_status_list" would not qualify, at least not without an explanation). In short, I'd really recommend bringing in someone from TortoiseGit, as they're more likely to know the answer to those questions (maybe they're in the middle of porting over, and that cache still has a `git`-y feeling to it?).

**Comment by ethomson** [comment]
Right - calling `git_status_file` or `git_status_list` with a path filter (instead of `git_repository_hashfile`) would hit the index, and should be able to avoid reading the file from the working directory, performing CR/LF translation and hashing those results.  Using the status APIs, workdir files will be `stat`ed against the workdir cache and if they're up-to-date then the expensive bits will be skipped entirely.  I would expect that moving over to the status API this to be a much bigger perf win.

This is not to say that reading the same file several times is good - it's not - but since we're all strapped for time here, I want to try to get the big and easy wins first and then fix up the hard ones.

It might be worth bringing @csware into the conversation so that we can better understand the goals here and if TortoiseGit can be using `git_status_file`.

**Comment by mehrdadn** [comment]
If I can read between your lines correctly, you seem to think the slowness is from reading files and hashing them? You probably think that way because you're on MacOS. That is not the case. In fact you may notice the slowness above is for `GetFileAttributesEx`, not for `ReadFile`. On Windows the I/O subsystem is designed for efficiency around handles (i.e. descriptors), and *not* paths. It's designed such that merely accessing any file information by *paths* has a huge performance hit⁠—even if the file has zero length and you're not trying to read its data at all.

All of this is to say that, regardless of whether `git_status_list` is a suitable API here (I have no idea), what I'm trying to get across is that by hitting the file system 3x (if not more) more than you need to, you've made `git_repository_hashfile` 3x as slow (or slower) as it needs to be *in the use cases you intended it for*, and you're not going to see that factor drop except for pretty large files, where the I/O on the actual data would dominate. That ought to be enough for you to worry about regardless of whether a program called TortoiseGit exists in the world or not.

So *I* wasn't going to be one to drag in TortoiseGit folks because the slowness I was reporting here stands is an independent bug report, and even if they eliminated every call to `git_respository_hashfile` from their codebase you would still have exactly the same slowness on your hands that you do now; you just won't have the same user for that API to observe the slowness it for you. But I'm not actively opposed to it; the collaboration could be productive in other ways regardless. Just saying it's missing the point of *this* issue.

**Comment by mehrdadn** [comment]
The "provided FS trace" is only as big as my screen. My screen is only so big. That's why I said it's just a "snippet". I would've hoped that when you insisted on a repro, and when I spent 2 hours providing one, one of you would actually *use* it to reproduce the problem. And see for yourself that there's even more FS hits than what I have on the screen. Not just ignore it.

I don't understand anything about what you are saying. Somehow you're making this way more complicated than it actually is. The situation is very, very simple: there are *six* back-to-back calls just to CreateFile for `.git/index`. The FS is not being modified in between, so you can reduce this down to 1-2 calls. Similarly goes for any other file that you're reading more than once. This *will* speed things up for callers of that function, entirely independently of whether a program called "TortoiseGit" exists in this universe or not. That's all there is to it.

**Comment by mehrdadn** [comment]
P.S. I think it's insulting that this issue is still "Closed". It's like you're keeping the door shut in my face but still somehow finding it worthwhile to talk to me from behind a closed door. Why? Is it so hard to admit the discussion is worth having after all, and that closing it early happened far too prematurely? I don't see why I should be trying to help when I'm constantly being told the matter is closed.

**Comment by tiennou** [comment]
I've rewritten the test to be clar-based, and made a few runs on macOS (so take that with a grain of salt w.r.t reproducibility), sadly that's screenshot-heavy because Instruments lost its text export 😭.

<details><summary>Test case</summary><p>

```c
void test_repo_hashfile__fscache(void)
{
	git_buf contents = GIT_BUF_INIT_CONST("foo", strlen("foo"));
	git_repository *repo = NULL;
	git_index *index;
	git_oid tree_oid, commit_oid, file_oid;
	git_tree *tree;
	git_signature *sig;

	cl_git_pass(git_signature_new(&sig, "Hasher", "hasher@example.org", time(NULL), 0));

	cl_git_pass(git_repository_init(&repo, "hashfile", 0));
	cl_must_pass(p_chdir("hashfile"));
	cl_must_pass(git_futils_writebuffer(&contents, "a", O_CREAT|O_RDWR, GIT_FILEMODE_BLOB));

	cl_git_pass(git_repository_index(&index, repo));
	cl_git_pass(git_index_add_bypath(index, "a"));
	cl_git_pass(git_index_write_tree(&tree_oid, index));
	cl_git_pass(git_tree_lookup(&tree, repo, &tree_oid));

	cl_git_pass(git_commit_create(&commit_oid, repo, "refs/heads/master", sig, sig, NULL, "Initial commit", tree, 0, NULL));
	git_index_free(index);
	git_tree_free(tree);
	git_repository_free(repo);

	cl_git_pass(git_repository_open(&repo, "."));
	cl_git_pass(git_repository_hashfile(&file_oid, repo, "a", GIT_OBJECT_BLOB, NULL));

	git_repository_free(repo);
	git_signature_free(sig);
}
```

</p></details>

<details><summary>Filesystem activity during the test</summary><p>

<img width="800" alt="Capture d'écran 2019-06-19 12 01 13" src="https://user-images.githubusercontent.com/81099/59756470-0e9fb300-928a-11e9-966c-95482267b2f4.png">
</p></details>

<details><summary>Inverted call tree, focusing only calls by git_repository_hashfile</summary><p>

*Call-tree have been pruned at soon as the higher level system has been identified.*

<img width="615" alt="Capture d'écran 2019-06-19 12 20 33" src="https://user-images.githubusercontent.com/81099/59757756-c03fe380-928c-11e9-90ec-3c54c9f5e91a.png">
</p></details>

<details><summary>Event log with impacted paths</summary><p>

<img width="593" alt="Capture d'écran 2019-06-19 12 08 37" src="https://user-images.githubusercontent.com/81099/59756922-fbd9ae00-928a-11e9-9838-c6389eead39a.png">
</p></details>

<details><summary>Activity with the repo reused</summary><p>

*This should more closely match what is happening, as warm caches from the repo can be used to improve performance.*

<img width="617" alt="Capture d'écran 2019-06-19 11 53 59" src="https://user-images.githubusercontent.com/81099/59757062-36dbe180-928b-11e9-9c7b-b36bbc19bd73.png"></p></details>

<details><summary>Activity with core.autocrlf set</summary><p>

<img width="615" alt="Capture d'écran 2019-06-19 12 20 33" src="https://user-images.githubusercontent.com/81099/59759417-4f022f80-9290-11e9-8343-b5f7f7d082f6.png">
</p></details>

<hr>

I thus can't reproduce, as I can't see 1000th of calls `stat`ing ".git/index". The only thing I find strange in all those events are the somewhat noisy config freshness checks. Hence, macOS doesn't seem impacted.

Since I'm unaware of ProcMon's usage, I can only comment (and sorry for doubting you, I just want to confirm) that, *maybe* there was no filtering, and the `.git/index` writes are not coming from your test ? Alternatively, could the fact that Windows can have crlf conversion to do could cause more activity through the index ? I've since tested that, and it doesn't seem to matter, but it's imperfect, since the tested file has no line endings.

**Comment by ethomson** [comment]
@tiennou if you're going to pursue this, could you open a new issue for it that describes the problem you're trying to solve so that we can track it?  This issue both different problem (`core.fscache`) and has turned into a more meta-discussion which would be more appropriate in https://github.com/libgit2/discussion or irc, but not our bug tracker.  Thanks!

**Comment by mehrdadn** [comment]
@tiennou  I tested your code and it also resulted in a similar number of FS hits as my example did. It's not a filtering issue; I'm only looking at that process. And I tried `autocrlf = false` and it didn't seem to change anything. I would recommend you make sure you're finding all FS calls. If you're not missing anything, then somehow the behavior must be different on Mac than on Windows...

**Comment by mehrdadn** [comment]
@tiennou So here's one source of too many calls: in `git_futils_readbuffer_updated()` there is a call to `p_stat(path, &st)` and then a subsequent call to `git_futils_open_ro`. These hit the FS twice for the same file, when they could just do this only once.

**Comment by mehrdadn** [comment]
@tiennou Another source is of excessive calls is these lines in `git_index_set_caps`:

		if (!git_repository__cvar(&val, repo, GIT_CVAR_IGNORECASE))
			index->ignore_case = (val != 0);
		if (!git_repository__cvar(&val, repo, GIT_CVAR_FILEMODE))
			index->distrust_filemode = (val == 0);
		if (!git_repository__cvar(&val, repo, GIT_CVAR_SYMLINKS))
			index->no_symlinks = (val == 0);

Every single call to `git_repository__cvar` is attempting to hit `git_futils_readbuffer_updated()` three times, which itself attempts the (non-existent) local `.gitconfig` file twice, resulting in 6 accesses. These should really only be 1 access.

**Comment by mehrdadn** [comment]
Oops, fixed a typo^

**Comment by mehrdadn** [comment]
**Update:** It seems to be a CRLF issue. I have no idea what it means, but that's what this suggests, because it's repeatedly getting `GIT_CVAR_NOT_CACHED`:

![image](https://user-images.githubusercontent.com/523212/59764488-bac4a800-9250-11e9-9723-d323c26cb472.png)

**Comment by mehrdadn** [comment]
Or maybe it's not related to CRLF... I can't tell.

It seems `config_is_modified` is called *every single time* `git_config__lookup_entry` is called. I'm not sure that's necessary... is it?

**Comment by mehrdadn** [comment]
The vast majority of the FS hits seem to be these types of accesses to various config files, so reducing those would considerably speed things up. In principle `git_repository_hashfile` wouldn't need more than 1 access to each config file. However, I would further guess that it can reduce that to *zero* accesses on subsequent calls, so merely reducing it to 1 hit per call would still be very suboptimal.

All that aside, there are 5 accesses to `./hashfile/a`, which is, again, excessive. One is due to `git_path_isdir` calling `p_stat`, one is due to `p_realpath`, one is due to `git_path_isdir` *again*, one is due to `p_realpath` *again*, and one is due to `p_open`. I think it should be possible to reduce these to 1-2 calls as well.

**General note:** I would view filesystem hits as being as expensive as network accesses. You don't hit the network unless it's absolutely necessary, and the same should go for the file system. Not to mention the fact that these files could easily be on the network, in which case every single hit would be *really* slow.

**Comment by tiennou** [comment]
> I would recommend you make sure you're finding all FS calls.

That's dtrace-based, so *hopefully* it can be trusted (I do, at least).

> It seems config_is_modified is called every single time git_config__lookup_entry is called. I'm not sure that's necessary... is it?

Well, it's to make sure that, if the setting had been set externally, it's now taken into account. But you can't be sure without `stat`, or checking a hash (which is more costly but precise w.r.t timestamps). Ideally, we could *maybe* improve performance by skipping a few, but the larger stat-storms you're seeing are still not explained.

@ethomson
> if you're going to pursue this, could you open a new issue for it that describes the problem you're trying to solve so that we can track it?

I've done everything I could here, and can't reproduce on my platform, so sadly I'm done pursuing this. I'm well-aware that those stat-storms are weird, but haven't been able to form an explanation as to why they happen, and the 12 `stat`s *I* can see, though they could technically be brought down using caching, are required by the design of `git_repository_hashfile`.

**Comment by mehrdadn** [comment]
@tiennou 

> Well, it's to make sure that, [...]

I don't get it... if you're saying is *supposed* to happen, then how are you not seeing it on your end? What are you even _expecting_ to see? Maybe you're expecting to see the wrong thing?

> if the setting had been set externally, it's now taken into account.

This makes no sense. Within a **single** call to `git_repository_hashfile` you should need only 1 snapshot of the config file and reuse that, not re-read the file 3 times. That's just insane. If you can't even assume that the file is constant within a single `git_repository_hashfile` call then it's already game over. The file might well change in the *middle* of your read, and suddenly you'll get an inconsistent read. At some point you need to assume the world is sane. Or if you can't trust the user, you lock the file for exclusive access and start using descriptors/handles instead of paths. Either way, re-reading it 3+ times within a single call won't get you any more correctness than you already had.

> the larger stat-storms you're seeing are still not explained.

They are.... I just listed a *ton* of them for you above... and explained how the 1x is degenerating into 6x, etc....

**Comment by ethomson** [comment]
OK, I filed #5126 to track this particular issue.

To set expectations, this is not something I'll have time to look at in the immediate future.

**Comment by mehrdadn** [comment]
That's fine, I'm not in any rush. Thanks!

**Comment by mehrdadn** [comment]
@ethomson @tiennou In hindsight, it seems to me you got the impression that this whole argument was just over config and index files. That wasn't the case... this issue was far broader than that. My example stack traces were just *examples*—they couldn't possibly show every poor access pattern for every file. That's why I tried so hard to get you guys to repro this yourselves... so you could see this is an issue all over the library. This library has excessive FS hits for pretty much *every* file, and `config`/`index` files are not necessarily the worst offenders. I listed more examples on #5126 that appear to result in excessive hits even on *subsequent* calls—and again, I don't claim to have an exhaustive list. You *will* have to look at the ProcMon traces yourselves to eliminate all the wasteful calls.

**Comment by csware** [comment]
I just came home from vacation and now I see this long thread.

> The only thing I can tell is that TortoiseGit is explicitly asking us to read/filter the file, updating the odb/index to match what's on disk, because that's what git_repository_hashfile is meant to do.

That's is actually not correct. I neither expect `git_repository_hashfile` to update the odb/index nor it is documented in the link provided. We use this API because we expected that it just hashes a file with filters and returns the OID - nothing more.  Also, I expect that the internal cache for config and attr files/values are used and just need to be checked if the file on disk has changed (these checks should also be just conducted once per `git_repository_hashfile` call; we keep the git_repository struct alive as long as we know, that any config hasn't changed).

We're not using git_status on the whole working tree as we know which file has modified, we already have the index in memory (as well as the git_tree) and we don't need to see any patches generated. Also, TortoiseGit has logic to not hash files which exceed a certain size (that's is also not possible with git_status).

**Comment by mehrdadn** [comment]
@csware

> these checks should also be just conducted once per git_repository_hashfile call; we keep the git_repository struct alive as long as we know, that any config hasn't changed

A little confused regarding this -- if the git_repository stuct is alive as long as configs haven't changed, then the configs shouldn't need to be re-checked on every single call to git_repository_hashfile, right? Otherwise you'd have freed the struct in between calls.

**Comment by csware** [comment]
> A little confused regarding this -- if the git_repository stuct is alive as long as configs haven't changed, then the configs shouldn't need to be re-checked on every single call to git_repository_hashfile, right? Otherwise you'd have freed the struct in between calls.

In the TortoiseGit case the configs don't need to be rechecked, as this is already done in TortoiseGit. However, in the general case there could be hours/days between two `git_repository_hashfile` calls (disabling that would be nice, but it could be too specific to TGit). For .gitattribute files there is no detection code in TGit so far.

**Comment by mehrdadn** [comment]
Ah I see. Maybe they should add a flag to skip various checks (configs, attributes, etc.) so TortoiseGit can take advantage of that for whatever it's already monitoring?

**Comment by csware** [comment]
That could be an option (also, ignore files don't need to be checked at all for `git_repository_hashfile`, dunno if those are also read) - however, I think that might be too TortoiseGit specific.

(off-topic: @ethomson I'd like to discuss a very specific API for ignore file checking with you, at best via chat or voice/phone)

**Comment by mehrdadn** [comment]
One thing I'm still a little uncomfortable about is whether configs should be allowed to change when a repo is open. *So* many things could go wrong when a config file is allowed to change randomly. Like what if the user changes the `worktree` location? Does this library really support any change whatsoever? If not, is it documented what changes are supported? I would've expected the library to lock the repo and assume it's not changing in between calls...

**Comment by csware** [comment]
Locking the repository would just casue lots of issues (cf. https://gitlab.com/TortoiseGit/TortoiseGit/issues/497). In the TGitCache case, the library should just fail the operation.

**Comment by mehrdadn** [comment]
I think there's a "proper" fix for this, which is to open *everything* with `FILE_SHARE_DELETE` permissions, but it'd be quite an involved and platform-specific undertaking that I don't expect anyone would have an appetite for.

However... this seems only an issue when using TGitCache. For that, it seems reasonable to have a flag in `git_repository_open` to specify whether to lock the repo, so that a program like TGitCache can opt if it really wants to. (Or opt in, but I think opt-out makes more sense because there's just no way this library can possibly guarantee correct behavior in the face of random changes, so the caller should be waiving this explicitly.)

(It's probably worth mentioning, I personally never use TGitCache because I've run into so many issues like that one, and in any case I have no need for it either; the normal shell checks are still somewhat usable (still slow, but not outright unusable), and they would be more than fast enough if libgit2 was more efficient about things and didn't hit the file system all the time.)

**Comment by mehrdadn** [comment]
By the way, just some food for thought, but libgit2 could, hypothetically, allow the caller to specify custom functions for file system access. This would let a process like TGitCache incorporate its own cache of the file system into libgit2. Or even do fancier things, like using NTFS TxF to guarantee repo consistency with file system transactions. ;)

However, it'd be a fairly involved change, and a client like TortoiseGit would also need to be extremely careful performing any *writes* with this, since any discrepancy with the actual file system could lead to corruption. Mostly just wanted to mention it for completeness.
